src/gc/gc.cpp

```diff
@@ -36150,15 +36150,17 @@ unsigned int GCHeap::GetCondemnedGeneration()
     return gc_heap::settings.condemned_generation;
 }
 
-void GCHeap::GetMemoryInfo(uint32_t* highMemLoadThreshold, 
-                           uint64_t* totalPhysicalMem, 
-                           uint32_t* lastRecordedMemLoad,
+void GCHeap::GetMemoryInfo(uint64_t* highMemLoadThresholdBytes, 
+                           uint64_t* totalAvailableMemoryBytes,
+                           uint64_t* lastRecordedMemLoadBytes,
+                           uint32_t* lastRecordedMemLoadPct,
                            size_t* lastRecordedHeapSize,
                            size_t* lastRecordedFragmentation)
 {
-    *highMemLoadThreshold = gc_heap::high_memory_load_th;
-    *totalPhysicalMem = gc_heap::total_physical_mem;
-    *lastRecordedMemLoad = gc_heap::last_gc_memory_load;
+    *highMemLoadThresholdBytes = (uint64_t) (((double)gc_heap::high_memory_load_th) / 100 * gc_heap::total_physical_mem);
+    *totalAvailableMemoryBytes = gc_heap::heap_hard_limit != 0 ? gc_heap::heap_hard_limit : gc_heap::total_physical_mem;
+    *lastRecordedMemLoadBytes = (uint64_t) (((double)gc_heap::last_gc_memory_load) / 100 * gc_heap::total_physical_mem);
+    *lastRecordedMemLoadPct = gc_heap::last_gc_memory_load;
```

Comment:
Is the `last_gc_memory_load`  relative to `total_physical_mem`; or relative to `*totalAvailableMemoryBytes` ?

Either this or the previous line is wrong. When memory load is `100%`, `loadBytes` should be equal `totalAvailableMemoryBytes`.

It may be best to get rid of the `*lastRecordedMemLoadPct` argument, and just divide the two numbers on the GC side.

Reply:
last_gc_memory_load is relative to total_physical_mem. the problem is we do have a GetMemoryLoad that needs the actual memory load. 

I would really prefer our original API proposal that had both the total physical memory and the hardlimit for the GC heap 'cause we do care about both. and I don't think it's unreasonable for the users to know about both. either that or we have another API (internal) that gets the actual physical memory load.

Reply:
`last_gc_memory_load` is relative to `total_physical_mem`. If `totalAvailableMemoryBytes` is less than `total_physical_mem`, then it will be impossible for `last_gc_memory_load` to ever reach 100% -- it would be at most `totalAvailableMemoryBytes / total_physical_mem * 100`.

Reply:
The primary reason why we have introduced the `GCMemoryInfo` API is to allow folks to build self-tuning caches. 

Could you please review the places where this API is used currently (our `TlsOverPerCoreLockedStacksArrayPool` in particular - not sure whether there is anything else) and validate that it works as correctly in all cases (regular VMs, containers, hardlimit)?

The current code for this is here: https://github.com/dotnet/coreclr/blob/4ca032d49b0718eb40f2105f94db2b243cf94957/src/System.Private.CoreLib/shared/System/Buffers/TlsOverPerCoreLockedStacksArrayPool.cs#L303 . Is this code correct for all situations? If not, what is the code that we want people to write to implement this logic? Once we know what code we want people to write here, we should be clear about what these APIs needs to do.

> `GetMemoryLoad` that needs the actual memory load.

This is used by the internal experimental `MemoryLoadChangeNotification` APIs. Is this API used by anything?

Reply:
the impl in TlsOverPerCoreLockedStacksArrayPool.cs looks correct - it could be improved (for example it could say only trim if the size it trims could actually make a difference in memory pressure) but I don't think we are doing that for 3.0. 

it seems awkward to have to fit the 2 numbers (actual physical memory and hardlimit) into 1. and if we don't want to have to call the diagnostics API to retrieve the physical memory (which also seems awkward for GC) our original proposal seems more sensible to me.

Reply:
The impl in TlsOverPerCoreLockedStacksArrayPool does not take the hard limit into account at all with the current implementations of the APIs.

Let's say I have 32GB physical memory with 16GB of free memory, set `COMPlus_GCHeapHardLimit` to 200MB, and start using array pool heavily. The array pool is going to see HighMemoryLoadThresholdBytes=~30GB and MemoryLoadBytes=~16GB, so it will not make any attempt to trim and cache way over 200MB. Do you agree that this is the behavior today? It does not sound right to me. If I set 200MB hardlimit, the ArrayPool should try to help with staying under 200MB.

Reply:
it is the behavior today and yes it can be improved. I'm saying I don't think we have time to improve this in 3.0. and this is not a common scenario.

Reply:
So why are we introducing the HardLimit setting at all when it does not actually work end-to-end?

We should either make the HardLimit setting work end-to-end; or pull it from 3.0 (it would be ok to keep it as "experimental feature" that does not fully work).

Reply:
because hardlimit was needed it for the container scenario where previously it worked very poorly. we can improve the usage in TlsOverPerCoreLockedStacksArrayPool in the future releases. I have not heard anyone complain about TlsOverPerCoreLockedStacksArrayPool not working.

Reply:
Also, we need to have a good forward looking example for how basic cache self-tuning should be written using the GetGCMemory APIs. We should avoid changing this example every release - doing so would cause a lot of confusion.

Reply:
I'd like that too; we didn't have enough time to do it in this release, which was unfortunate.

Reply:
What logic do we think would need to look at physical memory rather than the hard limit?

Reply:
We have received a lot of feedback in 2.x about ArrayPool keeping too much memory over time. We have seen people taking extreme measures like poking the ArrayPool via private reflection to make it release the memory. We have implemented the trimming at the last minute to mitigate the situation. 

If the ArrayPool trimming does not take the hardlimit into account, I expect that we are going to hear the same feedback again from folks who use the hardlimit.

Reply:
Right I'm agreeing with your assessment. I'm mostly wondering when you'd need the actual physical memory vs the hard limit (the other should probably be exposed where we have the other process specific APIs Process.*)

Reply:
(My reply was for Maoni's comment.)

I agree that the actual physical memory does not seem to be necessary to build decent trimming heuristic.

I think that there are 4 numbers to feed into a decent basic trimming heuristic:
- The GC heap size: current value and limit that it should stay under. I believe it is `HeapSizeBytes` and `TotalAvailableMemoryBytes` in the current API.
- The memory load (ie the GC heap size plus everything else): current value and limit that it should stay under. I believe it is `MemoryLoadBytes` and `HighMemoryLoadThresholdBytes` in the current API.

When either of these gets too close to its limit, it is a signal to trim caches.

Reply:
I have opened https://github.com/dotnet/coreclr/issues/25552 on ArrayPool not respecting GCHardLimit.